### PR TITLE
ENH: Improve error message for empty object array

### LINKED ISSUE #23572: df.to_stata fails when a column of type object contains only None
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
df = pd.DataFrame({'a': ['a', None]})
df.to_stata('test.dta')
df = pd.DataFrame({'a': [None, 'a']})
df.to_stata('test.dta')
df = pd.DataFrame({'a': [None, None]})
df.to_stata('test.dta')
# ValueError: Writing general object arrays is not supported
```

#### Problem description

The `df.to_stata()` method writes columns containing `None` without error when there is at least one string value in the column, but fails if the column contains only `None`. It's unclear what data type to write a column of `None` as, so maybe that's why this isn't supported? I would propose that a column with values of only `None` be written as `str1` with empty strings.

I came across this error because I read in a Parquet file with `pd.read_parquet()` and was unable to write the file to Stata format. In the Parquet schema, the column had type `BYTE_ARRAY UTF8`, but since the column had only missing values, it was read into Pandas as only `None`.

#### Expected Output

Stata file written to disk with missing values for the column with `None`.

#### Output of ``pd.show_versions()``

<details>

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Linux
OS-release: 4.15.0-38-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: 3.3.2
pip: 18.0
setuptools: 40.0.0
Cython: 0.28.3
numpy: 1.14.2
scipy: 1.0.0
pyarrow: 0.11.1
xarray: None
IPython: 6.5.0
sphinx: 1.6.6
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2017.3
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.2
openpyxl: 2.4.10
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.1.14
pymysql: None
psycopg2: 2.7.5 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: 0.1.5
pandas_gbq: None
pandas_datareader: None
```

</details>

### PR DESCRIPTION
**Pull request by bashtage**
Improve the error message shown when an object array is empty

closes #23572

- [x] closes #23572
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

No what's new since too small a change and no test needed.

### CODE REVIEW

File: pandas/io/stata.py [hunk 1]
```diff
@@ -1868,7 +1868,12 @@ def _dtype_to_default_stata_fmt(dtype, column, dta_version=114,
         inferred_dtype = infer_dtype(column.dropna())
         if not (inferred_dtype in ('string', 'unicode') or
                 len(column) == 0):
-            raise ValueError('Writing general object arrays is not supported')
+            raise ValueError('Only string-like object arrays containing all '
+                             'strings or a mix of strings and None can be '
+                             'exported. Object arrays containing only null '
+                             'values are prohibited. Other object types'
+                             'cannot be exported and must first be converted '
+                             'to one of the supported types.')
```
Comment (gfyoung) [inline pandas/io/stata.py]: * By one of the supported types, do you just mean string?
* I presume a test for this error message is coming?

**Comment by bashtage** [comment]
I also extended the test in my previous PR to ensure that the case with all `None` when specifically requested to write as a strl passes.  I hope this is OK, but could move to a new PR is needed.

**Comment by jtkiley** [comment]
@bashtage Consider adding the offending column to the error like in #23646. The all `None` case isn't hard to handle, but most of the work for me is identifying the column(s) that are the problem in datasets with hundreds of columns.

Since it's throwing on a particular column, it may be that the first encountered column is the only one that can be named, but that's still helpful. An alternative would be providing a brief code example that will identify these all `None` columns (and perhaps an example of fixing it with `convert_strl`).

I could imagine relatively novice pandas users (i.e. Stata switchers) hitting this error, so I'd lean toward a little more hand-holding/direction. Happy to hear counterarguments, though.

**Comment by bashtage** [comment]
@jtkiley I added this.

File: pandas/io/stata.py (hunk 1 above)
Comment (bashtage) [reply pandas/io/stata.py]: > By one of the supported types, do you just mean string?

It could be any supported type, int8-32, float, double, string or datetime. The user has to make a choice as to what best expresses the object.

**Comment by jtkiley** [comment]
@bashtage Thanks for this (looks great) and all you do on the Stata format. It's a big help for those of us working with both, and it helps smooth the work I do to train Stata users to use pandas.

**Comment by jreback** [comment]
thanks @bashtage
